os/arch/arm/src/s5j/s5j_i2s.c

```diff
@@ -665,12 +665,10 @@ static int i2s_rxdma_setup(struct s5j_i2s_s *priv)
 	dmatask = bfcontainer->dmatask;
 	/* No data received yet */
 
-	apb->nbytes = 0;
-	apb->curbyte = 0;
 
-	dmatask->dst = (void *)apb->samp;
+	dmatask->dst = (void *)apb->samp + apb->curbyte;
```

Comment:
1. Will header be added in first buffer alone?
2. Why can't upper layer just have this details in its struct locally and use it when ever it is needed?
3. How will this handle if the sample /bit rates are changed on the fly?

Reply:
1. Yes it will be in first header alone
2. We need to maintain extra state there, here the change proposes that dma transfer respects the curbyte and nbytes field in apb buffer being passed. It just zeros them out and starts  transferring to the start of buffer, instead it can start transferring at an offset from start.
3. If this is not there , then it would be difficult to handle sample/bit rate changes on the fly since dma would overwrite every time dma transfer is done. The change is not handling any header construction logic. That is purely done in tinyalsa layer. It is just the context this change has come up.

Reply:
Now, after paying enough attention to the change (rather than the logic), will this code work the same all the time?
Coz, I dont know how compiler will decide on the bytes to increment  a void* pointer. 
Could you check if this code follows/obeys C standard?

Reply:
I think it would be safe to use char* cast to increment by apb->curbyte and then cast whole to (void*)

Reply:
@shivgarg Kindly do more research on deciding the char* type cast.  Further, one more question. If configured to record for 1 sec with one single buffer for 64kB at 16K X 2 bytes per sample X 2 channels, the receiver will receive 44 bytes less. Right? with real samples of 44 less and 44 bytes header. Does pcm_readi API description says this?

Reply:
char* should work fine. uintptr_t is defined in stdint.h as int type for void* pointer. I will apply that and push the change.
Yes the user will recieve 44 less bytes. The pcm_readi is silent about this . We need to define the semantics of pcm_readi clearly. The user should allocate space for header it is implementaion and ask for more bytes to get full 1 sec of data.

Reply:
@kishore-sn Please give your comments regarding pcm_readi semantics. Which way should we go with regards to header.

Reply:
I just checked the tinycap utility provided by tinyalsa for audio record. In that I can see that the wav header is added to the start of the file by the application. The audio framework is not responsible for giving the header. When pcm_readi is called, only the audio data is returned. 
I have seen the same thing being done in other implementations like android which use ALSA. So, it would be better to modify our code to omit the WAV header completely. 
We could modify pcm_enqueuebuffer() as below: